app/src/main/java/de/westnordost/streetcomplete/data/osm/osmquests/SpatialCache.kt

```diff
@@ -0,0 +1,151 @@
+
+    // returns keys that were not put to cache because they are not in fully contained tiles
+    fun replaceAllInBBox(entries: Collection<Pair<T, LatLon>>, bbox: BoundingBox): List<T> {
+        val tiles = bbox.asListOfEnclosingTilePos()
+        val completelyContainedTiles = tiles.filter { it.asBoundingBox(
+            tileZoom).isCompletelyInside(bbox) }
+        if (tiles != completelyContainedTiles)
+            Log.i(TAG, "replacing tiles not completely contained in BBox")
+        val ignoredKeys = replaceAllInTiles(entries, completelyContainedTiles)
+
+        // trim only when bbox!
+        trim()
+        return ignoredKeys
+    }
+
+    // may add tiles, but does not call trim()
+    private fun replaceAllInTiles(entries: Collection<Pair<T, LatLon>>, tiles: Collection<TilePos>): List<T> {
+        // create / replace tiles
+        tiles.forEach { tile ->
+            byTile.remove(tile)?.let { byKey.keys.removeAll(it) }
+            byTile[tile] = hashSetOf()
+        }
+        if (tiles.size == 1) {
+            // shortcut: replace old tile
+            byTile[tiles.single()] = entries.unzip().first.toHashSet()
+            byKey.putAll(entries)
+            return emptyList()
+        } else {
+            // put only what is inside tiles
+            // and return what wasn't put
+            val ignoredKeys = mutableListOf<T>() // could also use sth like filter or mapNotNull instead of mutable list
+            val bboxByTile = tiles.associateWith { it.asBoundingBox(tileZoom) }
+            entries.forEach { (key, pos) ->
+                for (tile in tiles) {
+                    if (bboxByTile[tile]!!.contains(pos)) {
+                        byTile[tile]!!.add(key)
+                        byKey[key] = pos
+                        return@forEach
+                    }
+                }
+                ignoredKeys.add(key)
+            }
+            return ignoredKeys
+        }
+    }
+
+    fun get(bbox: BoundingBox): List<T> {
+        val requiredTiles = bbox.asListOfEnclosingTilePos()
+
+        val tilesToFetch = requiredTiles.filterNot { byTile.containsKey(it) }
+        if (tilesToFetch.isNotEmpty()) {
+            val newEntries = fetch(tilesToFetch.minTileRect()!!.asBoundingBox(tileZoom))
+            replaceAllInTiles(newEntries, tilesToFetch)
+        }
+
+        val keys = requiredTiles.flatMap { tile ->
+            if (tile.asBoundingBox(tileZoom).isCompletelyInside(bbox))
+                byTile[tile]!!
+            else
+                byTile[tile]!!.filter { byKey[it]!! in bbox }
+        }
+
+        // resize cache now
+        // don't do this right after putting new data, as this might remove tiles we need
+        trim()
+
+        return keys
+    }
+
+    fun trim(size: Int = maxTiles) {
+        if (byTile.size <= size) return
+
+        val removedKeys = hashSetOf<T>()
+        while (byTile.size > size) {
+            byTile.remove(byTile.keys.first())?.let { removedKeys.addAll(it) }
```

Comment:
Thinking about the behavior of the cache, I think this is not the behavior we want to have. 

This clears from the cache the tile that has first been put into the cache. But this is not what we want. We want the tile to be cleared that hasn't been **accessed** for the longest time.

E.g. the tile around your GPS position is probably the first one to be put into the cache. But it should not be first one to be cleared, but the very last one.

So, a `LinkedHashMap` doesn't seem to do the job here. Possible solutions:
- on every access of a tile, remove that tile from the LinkedHashMap and add it again, so it is (probably?) moved to the end. Needs testing if this is actually how the LinkedHashMap works
- use a LinkedHashSet and do the same on that (remove TilePos on access, then add it again), use that LinkedHashSet to find the TilePos that hasn't been accessed the longest
- maybe something else, e.g. just a LinkedList<TilePos> or a Map<TilePos, Int> (int is the priority/position)

Reply:
The `LinkedHashMap` is already ordered by access (put and get) thanks to the `true` in `LinkedHashMap<TilePos, HashSet<T>>((maxTiles/0.75).toInt(), 0.75f, true)`, see https://stackoverflow.com/questions/35838739/linkedhashmap-ordering

Reply:
ah, cool